File: zircon.jvm.examples/src/main/java/org/hexworks/zircon/examples/GameMockupExample.java
package org.hexworks.zircon.examples;

import org.hexworks.zircon.api.CP437TilesetResources;
import org.hexworks.zircon.api.ColorThemes;
import org.hexworks.zircon.api.Components;
import org.hexworks.zircon.api.SwingApplications;
import org.hexworks.zircon.api.application.AppConfig;
import org.hexworks.zircon.api.component.Button;
import org.hexworks.zircon.api.component.Label;
import org.hexworks.zircon.api.component.Panel;
import org.hexworks.zircon.api.component.*;
import org.hexworks.zircon.api.data.Position;
import org.hexworks.zircon.api.data.Size;
import org.hexworks.zircon.api.graphics.BoxType;
import org.hexworks.zircon.api.graphics.Symbols;
import org.hexworks.zircon.api.grid.TileGrid;
import org.hexworks.zircon.api.resource.TilesetResource;
import org.hexworks.zircon.api.screen.Screen;
import org.hexworks.zircon.api.uievent.UIEventResponse;

import java.awt.*;
import java.util.Arrays;

import static org.hexworks.zircon.api.ComponentDecorations.box;
import static org.hexworks.zircon.api.uievent.ComponentEventType.ACTIVATED;

public class GameMockupExample {

    private static final String MAIN_MENU_LABEL = "M A I N   M E N U";
    private static final String NEW_GAME_BUTTON_LABEL = "N E W   G A M E";
    private static final String OPTIONS_BUTTON_LABEL = "O P T I O N S";
    private static final String QUIT_BUTTON_LABEL = "Q U I T";
    private static final String DIFFICULTY_LABEL = "D I F F I C U L T Y";
    private static final String BACK_LABEL = Symbols.ARROW_LEFT + " B A C K";
    private static final String APPLY_LABEL = "A P P L Y";

    private static final String[] DIFFICULTIES = new String[]{"TINGLE", "ANXIETY", "HORROR"};

    private static final TilesetResource TILESET = CP437TilesetResources.rexPaint20x20();
    private static final int MAIN_MENU_PANEL_WIDTH = 25;
    private static final int MAIN_MENU_PANEL_HEIGHT = 10;
    private static final int PANEL_SPACING = 2;
    private static final ColorTheme THEME = ColorThemes.arc();

    public static void main(String[] args) {
        Dimension screenSize = Toolkit.getDefaultToolkit().getScreenSize();
        double columns = screenSize.getWidth() / TILESET.getWidth();
        double rows = screenSize.getHeight() / TILESET.getHeight();
        Size terminalSize = Size.create((int) columns, (int) rows);

        final TileGrid tileGrid = SwingApplications.startTileGrid(AppConfig.newBuilder()
                .withDefaultTileset(TILESET)
                .withSize(terminalSize)
                .withDebugMode(true)
                .fullScreen()
                .build());

        // ==========
        // MAIN MENU
        // ==========

        Screen mainMenuScreen = Screen.create(tileGrid);
        Position menuPosition = Position.create(
                (terminalSize.getWidth() - MAIN_MENU_PANEL_WIDTH) / 2,
                (terminalSize.getHeight() - MAIN_MENU_PANEL_HEIGHT) / 2);
        Label mainMenuLabel = Components.label()
                .withText(MAIN_MENU_LABEL)
                .withPosition(menuPosition.withRelativeY(-3).withRelativeX(4))
                .build();
        mainMenuScreen.addComponent(mainMenuLabel);

        Panel menuPanel = Components.panel()
                .withDecorations(box(BoxType.LEFT_RIGHT_DOUBLE))
                .withPosition(menuPosition)
                .withSize(Size.create(MAIN_MENU_PANEL_WIDTH, MAIN_MENU_PANEL_HEIGHT))
                .build();

        Button newGameButton = Components.button()
                .withText(NEW_GAME_BUTTON_LABEL)
                .withPosition(Position.create(3, 1))
                .build();
        menuPanel.addComponent(newGameButton);

        Button optionsButton = Components.button()
                .withText(OPTIONS_BUTTON_LABEL)
                .withPosition(Position.create(4, 3))
                .build();
        menuPanel.addComponent(optionsButton);

        Button quitButton = Components.button()
                .withText(QUIT_BUTTON_LABEL)
                .withPosition(Position.create(7, 5))
                .build();
        menuPanel.addComponent(quitButton);

        mainMenuScreen.addComponent(menuPanel);
        mainMenuScreen.setTheme(THEME);

        // ==========
        // OPTIONS
        // ==========

        Screen optionsScreen = Screen.create(tileGrid);

        Button backButton = Components.button()
                .withText(BACK_LABEL)
                .withPosition(Position.create(
                        PANEL_SPACING,
                        terminalSize.getHeight() - (PANEL_SPACING * 2)))
                .build();
        optionsScreen.addComponent(backButton);

        Button applyButton = Components.button()
                .withText(APPLY_LABEL)
                .withPosition((Position.create(PANEL_SPACING, 0)).relativeToRightOf(backButton))
                .build();
        optionsScreen.addComponent(applyButton);

        Panel difficultyPanel = Components.panel()
                .withSize(Size.create((terminalSize.getWidth() - PANEL_SPACING) / 3, 9))
                .withPosition(Position.create(PANEL_SPACING, PANEL_SPACING))
                .withDecorations(box(BoxType.LEFT_RIGHT_DOUBLE, DIFFICULTY_LABEL))
                .build();

        VBox difficultyBox = Components.vbox()
                .withSize(difficultyPanel.getSize().minus(Size.create(2, 2)))
                .build();

        RadioButtonGroup difficultyGroup = Components.radioButtonGroup().build();

        Arrays.asList(DIFFICULTIES).forEach((diff) -> {
            RadioButton btn = Components.radioButton()
                    .withText(diff)
                    .withKey(diff)
                    .build();
            difficultyBox.addComponent(btn);
            difficultyGroup.add(btn);
        });

        difficultyPanel.addComponent(difficultyBox);
        optionsScreen.addComponent(difficultyPanel);


        optionsScreen.setTheme(THEME);

        // INTERACTIONS

        quitButton.handleComponentEvents(ACTIVATED, (event) -> {
            System.exit(0);
            return UIEventResponse.processed();
        });

        optionsButton.handleComponentEvents(ACTIVATED, (event) -> {
            optionsScreen.display();
            return UIEventResponse.processed();
        });

        backButton.handleComponentEvents(ACTIVATED, (event) -> {
            mainMenuScreen.display();
            return UIEventResponse.processed();
        });

        // START IT UP
        mainMenuScreen.display();

    }
}

File: zircon.jvm.examples/src/main/kotlin/org/hexworks/zircon/examples/other/GameMockupExample.kt
package org.hexworks.zircon.examples.other

import org.hexworks.zircon.api.CP437TilesetResources.rexPaint20x20
import org.hexworks.zircon.api.ColorThemes.arc
import org.hexworks.zircon.api.ComponentDecorations.box
import org.hexworks.zircon.api.Components.button
import org.hexworks.zircon.api.Components.label
import org.hexworks.zircon.api.Components.panel
import org.hexworks.zircon.api.Components.radioButton
import org.hexworks.zircon.api.Components.radioButtonGroup
import org.hexworks.zircon.api.Components.vbox
import org.hexworks.zircon.api.SwingApplications.startTileGrid
import org.hexworks.zircon.api.application.AppConfig.Companion.newBuilder
import org.hexworks.zircon.api.data.Position
import org.hexworks.zircon.api.data.Size
import org.hexworks.zircon.api.graphics.BoxType
import org.hexworks.zircon.api.graphics.Symbols
import org.hexworks.zircon.api.screen.Screen.Companion.create
import org.hexworks.zircon.api.uievent.ComponentEvent
import org.hexworks.zircon.api.uievent.ComponentEventType
import org.hexworks.zircon.api.uievent.UIEventResponse.Companion.processed
import java.awt.Toolkit
import java.util.*
import java.util.function.Consumer

object GameMockupExample {

    private val MAIN_MENU_LABEL = "M A I N   M E N U"
    private val NEW_GAME_BUTTON_LABEL = "N E W   G A M E"
    private val OPTIONS_BUTTON_LABEL = "O P T I O N S"
    private val QUIT_BUTTON_LABEL = "Q U I T"
    private val DIFFICULTY_LABEL = "D I F F I C U L T Y"
    private val BACK_LABEL = Symbols.ARROW_LEFT.toString() + " B A C K"
    private val APPLY_LABEL = "A P P L Y"

    private val DIFFICULTIES = arrayOf("TINGLE", "ANXIETY", "HORROR")

    private val TILESET = rexPaint20x20()
    private val MAIN_MENU_PANEL_WIDTH = 25
    private val MAIN_MENU_PANEL_HEIGHT = 10
    private val PANEL_SPACING = 2
    private val THEME = arc()

    @JvmStatic
    fun main(args: Array<String>) {
        val screenSize = Toolkit.getDefaultToolkit().screenSize
        val columns = screenSize.getWidth() / TILESET.width
        val rows = screenSize.getHeight() / TILESET.height
        val gridSize = Size.create(columns.toInt(), rows.toInt())
        val tileGrid = startTileGrid(
            newBuilder()
                .withDefaultTileset(TILESET)
                .withSize(gridSize)
                .withDebugMode(true)
                .withFullScreen(true)
                .build()
        )

        // ==========
        // MAIN MENU
        // ==========
        val mainMenuScreen = create(tileGrid)
        val menuPosition = Position.create(
            (gridSize.width - MAIN_MENU_PANEL_WIDTH) / 2,
            (gridSize.height - MAIN_MENU_PANEL_HEIGHT) / 2
        )
        val mainMenuLabel = label()
            .withText(MAIN_MENU_LABEL)
            .withPosition(menuPosition.withRelativeY(-3).withRelativeX(4))
            .build()
        mainMenuScreen.addComponent(mainMenuLabel)
        val menuPanel = panel()
            .withDecorations(box(BoxType.LEFT_RIGHT_DOUBLE))
            .withPosition(menuPosition)
            .withPreferredSize(Size.create(MAIN_MENU_PANEL_WIDTH, MAIN_MENU_PANEL_HEIGHT))
            .build()
        val newGameButton = button()
            .withText(NEW_GAME_BUTTON_LABEL)
            .withPosition(Position.create(3, 1))
            .build()
        menuPanel.addComponent(newGameButton)
        val optionsButton = button()
            .withText(OPTIONS_BUTTON_LABEL)
            .withPosition(Position.create(4, 3))
            .build()
        menuPanel.addComponent(optionsButton)
        val quitButton = button()
            .withText(QUIT_BUTTON_LABEL)
            .withPosition(Position.create(7, 5))
            .build()
        menuPanel.addComponent(quitButton)
        mainMenuScreen.addComponent(menuPanel)
        mainMenuScreen.theme = THEME

        // ==========
        // OPTIONS
        // ==========
        val optionsScreen = create(tileGrid)
        val backButton = button()
            .withText(BACK_LABEL)
            .withPosition(
                Position.create(
                    PANEL_SPACING,
                    gridSize.height - PANEL_SPACING * 2
                )
            )
            .build()
        optionsScreen.addComponent(backButton)
        val applyButton = button()
            .withText(APPLY_LABEL)
            .withPosition(Position.create(PANEL_SPACING, 0).relativeToRightOf(backButton))
            .build()
        optionsScreen.addComponent(applyButton)
        val difficultyPanel = panel()
            .withPreferredSize(Size.create((gridSize.width - PANEL_SPACING) / 3, 9))
            .withPosition(Position.create(PANEL_SPACING, PANEL_SPACING))
            .withDecorations(box(BoxType.LEFT_RIGHT_DOUBLE, DIFFICULTY_LABEL))
            .build()
        val difficultyBox = vbox()
            .withPreferredSize(difficultyPanel.size.minus(Size.create(2, 2)))
            .build()
        val difficultyGroup = radioButtonGroup().build()
        Arrays.asList(*DIFFICULTIES).forEach(Consumer { diff: String? ->
            val btn = radioButton()
                .withText(diff!!)
                .withKey(diff)
                .build()
            difficultyBox.addComponent(btn)
            difficultyGroup.addComponent(btn)
        })
        difficultyPanel.addComponent(difficultyBox)
        optionsScreen.addComponent(difficultyPanel)
        optionsScreen.theme = THEME

        // INTERACTIONS
        quitButton.handleComponentEvents(ComponentEventType.ACTIVATED) { event: ComponentEvent? ->
            System.exit(0)
            processed()
        }
        optionsButton.handleComponentEvents(ComponentEventType.ACTIVATED) { event: ComponentEvent? ->
            optionsScreen.display()
            processed()
        }
        backButton.handleComponentEvents(ComponentEventType.ACTIVATED) { event: ComponentEvent? ->
            mainMenuScreen.display()
            processed()
        }

        // START IT UP
        mainMenuScreen.display()
    }
}